model/waterfall.go

```diff
@@ -0,0 +1,618 @@
+	DisplayName      string                  `json:"display_name"`
+	TimeTaken        time.Duration           `json:"time_taken"`
+	Activated        bool                    `json:"activated"`
+	FailedTestNames  []string                `json:"failed_test_names,omitempty"`
+	ExpectedDuration time.Duration           `json:"expected_duration,omitempty"`
+	StartTime        int64                   `json:"start_time"`
+}
+
+// failedTest holds all the information for displaying context about tests that failed in a
+// waterfall page tooltip.
+
+// WaterfallVersion holds the waterfall UI representation of a single version (column)
+// If the RolledUp field is false, then it contains information about
+// a single version and the metadata fields will be of length 1.
+// If the RolledUp field is true, this represents multiple inactive versions, with each element
+// in the metadata arrays corresponding to one inactive version,
+// ordered from most recent inactive version to earliest.
+type WaterfallVersion struct {
+
+	// whether or not the version element actually consists of multiple inactive
+	// versions rolled up into one
+	RolledUp bool `json:"rolled_up"`
+
+	// metadata about the enclosed versions.  if this version does not consist
+	// of multiple rolled-up versions, these will each only have length 1
+	Ids                 []string    `json:"ids"`
+	Messages            []string    `json:"messages"`
+	Authors             []string    `json:"authors"`
+	CreateTimes         []time.Time `json:"create_times"`
+	Revisions           []string    `json:"revisions"`
+	RevisionOrderNumber int         `json:"revision_order"`
+
+	// used to hold any errors that were found in creating the version
+	Errors   []waterfallVersionError `json:"errors"`
+	Warnings []waterfallVersionError `json:"warnings"`
+	Ignoreds []bool                  `json:"ignoreds"`
+}
+
+// waterfallVersions implements the sort interface to allow backend sorting.
+type waterfallVersions []WaterfallVersion
+
+func (wfv waterfallVersions) Len() int {
+	return len(wfv)
+}
+
+func (wfv waterfallVersions) Less(i, j int) bool {
+	return wfv[i].RevisionOrderNumber > wfv[j].RevisionOrderNumber
+}
+
+func (wfv waterfallVersions) Swap(i, j int) {
+	wfv[i], wfv[j] = wfv[j], wfv[i]
+}
+
+type waterfallVersionError struct {
+	Messages []string `json:"messages"`
+}
+
+// waterfallBuildVariants implements the sort interface to allow backend sorting.
+type waterfallBuildVariants []waterfallBuildVariant
+
+func (wfbv waterfallBuildVariants) Len() int {
+	return len(wfbv)
+}
+
+func (wfbv waterfallBuildVariants) Less(i, j int) bool {
+	return wfbv[i].DisplayName < wfbv[j].DisplayName
+}
+
+func (wfbv waterfallBuildVariants) Swap(i, j int) {
+	wfbv[i], wfbv[j] = wfbv[j], wfbv[i]
+}
+
+// For given build variant, variant display name and variant search query
+// checks if matched variant has active tasks
+func variantHasActiveTasks(
+	b build.Build, bvDisplayName string, variantQuery string,
+) bool {
```

Comment:
can we avoid splitting the signature on multiple lines?

Reply:
Can you share a recommended codestyle for the function declarations? What is the string length at which I should separate it to multiple lines? For the two-panel editor on regular display and with the font of convenient size it is usually less then 100 symbols per line - that's what I think.